### PR TITLE
Tasty reflect symbol refactorings

### PR DESCRIPTION
**Pull request by nicolasstucki**
* Remove subtypes of Symbol
* Remove logic duplication
* Remove unnecessary Options from API

### CODE REVIEW

**Comment by nicolasstucki** [comment]
@milessabin this PR will update a file in shapeless

**Comment by milessabin** [comment]
@nicolasstucki thanks for the heads up!

**Comment by smarter** [comment]
With these simplifications, are we now at the point where we can use provide `ClassTag` instances for all the abstract types such that `ClassTag.unapply` can be used for type tests ? I think that was problematic previously because we needed to pass an implicit Context to the `unapply` method, but if the tasty-reflect AST closely matches the dotty AST that shouldn't be needed, since the implementation should just need to do instanceof checks.

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 1]
```diff
@@ -1458,6 +1458,16 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
 
   def Symbol_flags(self: Symbol)(given Context): Flags = self.flags
 
+  def Symbol_tree(self: Symbol)(given Context): Tree = {
+    assert(!self.is(Case))
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: A case class or case enum may also have the flag.

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 2]
```diff
@@ -1506,148 +1516,84 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
...
-
-  def PackageDefSymbol_tree(self: PackageDefSymbol)(given Context): PackageDef =
-    FromSymbol.packageDefFromSym(self)
-
-  type TypeSymbol = core.Symbols.TypeSymbol
-
-  def matchTypeSymbol(symbol: Symbol)(given Context): Option[TypeSymbol] =
-    if (symbol.isType) Some(symbol.asType) else None
+  def Symbol_isPackageDefSymbol(symbol: Symbol)(given Context): Boolean =
+    symbol.is(core.Flags.Package)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: `isPackageDefSymbol` -> `isPackage`?

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 3]
```diff
@@ -1506,148 +1516,84 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
...
 
-  type ClassDefSymbol = core.Symbols.ClassSymbol
+  def Symbol_isTypeSymbol(symbol: Symbol)(given Context): Boolean =
+    symbol.isType
 
-  def matchClassDefSymbol(symbol: Symbol)(given Context): Option[ClassDefSymbol] =
-    if (symbol.isClass) Some(symbol.asClass) else None
+  def Symbol_isClassDefSymbol(symbol: Symbol)(given Context): Boolean =
+    symbol.isClass
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: `isClassDefSymbol` -> `isClass`?

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 4]
```diff
@@ -1506,148 +1516,84 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
...
-    FromSymbol.typeBindFromSym(self)
-
-  type TermSymbol = core.Symbols.TermSymbol
-
-  def matchTermSymbol(symbol: Symbol)(given Context): Option[TermSymbol] =
-    if (symbol.isTerm) Some(symbol.asTerm) else None
-
-  type DefDefSymbol = core.Symbols.TermSymbol
+  def Symbol_isTypeBindSymbol(symbol: Symbol)(given Context): Boolean =
+    symbol.isType && symbol.is(core.Flags.Case)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: `isTypeBindSymbol` -> `isPatternBoundType`?

A case class symbol will pass the check.

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 5]
```diff
@@ -1506,148 +1516,84 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
...
 
-  def matchDefDefSymbol(symbol: Symbol)(given Context): Option[DefDefSymbol] =
-    if (symbol.isTerm && symbol.is(core.Flags.Method)) Some(symbol.asTerm) else None
+  def Symbol_isTermSymbol(symbol: Symbol)(given Context): Boolean =
+    symbol.isTerm
 
-  def DefDefSymbol_tree(self: DefDefSymbol)(given Context): DefDef =
-    FromSymbol.defDefFromSym(self)
+  def Symbol_isDefDefSymbol(symbol: Symbol)(given Context): Boolean =
+    symbol.isTerm && symbol.is(core.Flags.Method)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: `isDefDefSymbol` -> `isMethod`?

As discussed offline, normally only term symbols have the flag `Method`, thus there is no need to test `symbol.isTerm`.

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 6]
```diff
@@ -1506,148 +1516,84 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
...
-  }
-
-  def ValDefSymbol_companionClass(self: ValDefSymbol)(given Context): Option[ClassDefSymbol] = {
-    val sym = self.companionClass
-    if (sym.exists) Some(sym.asClass) else None
-  }
-
-  type BindSymbol = core.Symbols.TermSymbol
+  def Symbol_isValDefSymbol(symbol: Symbol)(given Context): Boolean =
+    symbol.isTerm && !symbol.is(core.Flags.Method) && !symbol.is(core.Flags.Case)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: An enum case will be fail the test. The method `Symbol.isPatternBound`:

```Scala
    final def isPatternBound(implicit ctx: Context): Boolean =
      !isClass && this.is(Case, butNot = Enum | Module)
```

File: library/src/scala/tasty/reflect/SymbolOps.scala [hunk 7]
```diff
@@ -83,181 +63,63 @@ trait SymbolOps extends Core {
...
+    def isType(given ctx: Context): Boolean = internal.Symbol_isTypeSymbol(self)
+    def isTerm(given ctx: Context): Boolean = internal.Symbol_isTermSymbol(self)
+    def isValDef(given ctx: Context): Boolean = internal.Symbol_isValDefSymbol(self)
+    def isDefDef(given ctx: Context): Boolean = internal.Symbol_isDefDefSymbol(self)
+    def isTypeDef(given ctx: Context): Boolean = self.isType && !self.flags.is(Flags.Case)
+    def isClassDef(given ctx: Context): Boolean = internal.Symbol_isClassDefSymbol(self)
+    def isBind(given ctx: Context): Boolean = self.isTerm && self.flags.is(Flags.Case)
+    def isTypeBind(given ctx: Context): Boolean = self.isType && self.flags.is(Flags.Case)
+    def isPackageDef(given ctx: Context): Boolean = internal.Symbol_isPackageDefSymbol(self)
+    def isNoSymbol(given ctx: Context): Boolean = self == Symbol.noSymbol
```
Comment (liufengyun) [inline library/src/scala/tasty/reflect/SymbolOps.scala]: `isDefDef ` -> `isMethod`?
`isTypeDef` -> `isType`?
`isClassDef` -> `isClass`?
`isBind` -> `isPatternBoundTerm`?
`isTypeBind` -> `isPatternBoundType`?
`isPackageDef` -> `isPackage`?
`isNoSymbol` -> `exists`?

**Comment by nicolasstucki** [comment]
> With these simplifications, are we now at the point where we can provide `ClassTag` instances for all the abstract types such that `ClassTag.unapply` can be used for type tests ? I think that was problematic previously because we needed to pass an implicit Context to the `unapply` method, but if the tasty-reflect AST closely matches the dotty AST that shouldn't be needed, since the implementation should just need to do instanceof checks.

No. There are still problematic ones on `Tree`, `Pattern` and `Id`.

**Comment by smarter** [comment]
OK, alternatively we could consider using `Typeable`, apparently the Typelevel Scala compiler was able to use this typeclass to `case x: T`: https://github.com/milessabin/shapeless/blob/bd3110d9e075b28ee1323d861e0aca130c6891b4/core/src/main/scala/shapeless/typeable.scala#L49 It seems more promising than ClassTag because the unapply method is in the companion object and looks like this:
```scala
def unapply[T: Typeable](t: Any): Option[T] = t.cast[T]
```
Notice that it delegates the unapply to the Typeable instance, so we could do somethng like this I think:
```scala
given patternIsTypeable(given Context): Typeable[Pattern] {
  def cast(t: Any): Option[T] = // ... operation that needs a Context
}
```

@milessabin WDYT ? That would require getting Typeable into the standard library. It'd be nice if we could use that to deprecate ClassTag-based pattern matching since this is known to be unsound: https://gist.github.com/Blaisorblade/a0eebb6a4f35344e48c4c60dc2a14ce6

File: library/src/scala/tasty/reflect/SymbolOps.scala (hunk 7 above)
Comment (nicolasstucki) [reply library/src/scala/tasty/reflect/SymbolOps.scala]: We already have an `isType. Also, those are supposed to have the names of the trees they represent.

Comment (liufengyun) [reply library/src/scala/tasty/reflect/SymbolOps.scala]: Thank for clarification, then it's better to leave as it is.

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: `Symbol_isClassDef`

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: `isPackageDef`

File: library/src/scala/tasty/reflect/SymbolOps.scala (hunk 7 above)
Comment (nicolasstucki) [reply library/src/scala/tasty/reflect/SymbolOps.scala]: I added documentation to show clearly the relation between the symbols and the trees.

File: compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala [hunk 8]
```diff
@@ -1572,20 +1563,33 @@ class ReflectionCompilerInterface(val rootContext: core.Contexts.Context) extend
...
+    symbol.is(Flags.Package)
+
+  def Symbol_isClassDef(symbol: Symbol)(given Context): Boolean =
+    symbol.isClass
+
+  def Symbol_isTypeDef(symbol: Symbol)(given ctx: Context): Boolean =
+    symbol.isType && !symbol.isClass && !symbol.is(Flags.Case)
+
+  def Symbol_isValDef(symbol: Symbol)(given Context): Boolean =
+    symbol.isTerm && !symbol.is(core.Flags.Method) && !symbol.is(core.Flags.Case)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/tastyreflect/ReflectionCompilerInterface.scala]: `!symbol.is(core.Flags.Case)` may need the same test for `Enum | Module`.

**Pull request merged**